Standalone signaling support

fancycode opened the pull request:
This PR adds support for using a standalone signaling server.

See https://github.com/nextcloud/spreed/wiki/Spreed-Signaling-API for the documentation of message formats. In a follow-up PR, the same messages will be implemented for the PHP backend, so there is only one API that clients need to implement.

@nickvergessen @Ivansss

fancycode commented:
@nickvergessen @Ivansss this is now ready for some testing :smiley:

fancycode commented:
@nickvergessen @Ivansss this is now pending for quite some time - would be great to get some feedback...

nickvergessen commented:
I will have a look on monday/tuesday, sorry conf was too busy

nickvergessen commented on lib/Controller/AppSettingsController.php:
```diff
@@ -99,6 +102,20 @@ public function setSpreedSettings($stun_server, $turn_server, $turn_server_secre
+				if (!filter_var($server, FILTER_VALIDATE_URL)) {
+					return array('data' =>
+						array('message' =>
+							(string) $this->l10n->t('Invalid signaling server url.')
```
cast is not required anymore

fancycode replied:
For consistency I would rather keep this aligned with surrounding code and remove the cast here and in other locations in a follow-up PR.

nickvergessen replied:
Yeah, same as in the comment below, just ignore for now ;)

nickvergessen commented on lib/Controller/AppSettingsController.php:
```diff
@@ -99,6 +102,20 @@ public function setSpreedSettings($stun_server, $turn_server, $turn_server_secre
+			$servers = explode('\n', $signaling_server);
+			foreach($servers as $server) {
+				if (!filter_var($server, FILTER_VALIDATE_URL)) {
+					return array('data' =>
```
you can use the short array syntax `[ ]`

nickvergessen replied:
Ah, just saw you only copy pasted, so okay

nickvergessen commented on lib/Controller/AppSettingsController.php:
```diff
@@ -55,8 +55,11 @@ public function __construct($appName,
 	 * @param string $turn_server_secret
+	 * @param string $signaling_server
+	 * @param string $signaling_secret
+	 * @param string $signaling_skip_verify_cert
```
order is mixed up
can you also add `$turn_server_protocols` to the docs?

fancycode replied:
Changed

nickvergessen commented on lib/Controller/AppSettingsController.php:
```diff
@@ -120,6 +137,24 @@ public function setSpreedSettings($stun_server, $turn_server, $turn_server_secre
+		}
+
+		$currentSignalingSkipVerifyCert = $this->config->getAppValue('spreed', 'signaling_skip_verify_cert', '');
+		if ($signaling_skip_verify_cert === null) {
```
```
$signaling_skip_verify_cert = (string) $signaling_skip_verify_cert;
```

fancycode replied:
Sorry, I don't understand what you want to have changed here.

nickvergessen replied:
You set the variable to an empty string when it is null, so you can also just cast it to string, getting rid of the if-block. Should have the same result?

fancycode replied:
php magic :see_no_evil:

nickvergessen commented on lib/Controller/RoomController.php:
```diff
@@ -305,6 +310,11 @@ protected function createOneToOneRoom($targetUserName) {
 			$room->addParticipant($targetUser->getUID(), Participant::OWNER);
 			$this->createNotification($currentUser, $targetUser, $room);
 
+			$this->backend->roomInvited($room, [
```
I think I would move this into the methods of the `Room` class.
For better decoupling we could also trigger a hook? This would allow any app to listen to those events and provide a signaling backend?

fancycode replied:
This is now updated to use events in the latest commit.

nickvergessen commented on lib/Controller/SignallingController.php:
```diff
@@ -76,6 +82,11 @@ public function __construct($appName,
 	public function signalling($messages) {
+		$signaling = $this->config->getSignalingServer();
+		if (!empty($signaling)) {
+			throw new \Exception('Internal signaling disabled.');
```
Should `return new JSONResponse('Internal signaling disabled.', Http::STATUS_BAD_REQUEST);` instead of throwing?

fancycode replied:
Changed

nickvergessen commented on lib/Controller/SignallingController.php:
```diff
@@ -259,4 +251,149 @@ protected function getUsersInRoom(Room $room) {
+	 * @return JSONResponse
+	 */
+	public function backend() {
+		$json = file_get_contents('php://input');
```
can't you use `$this->request` ?

fancycode replied:
`$this->request` is a `OCP\IRequest` which only provides access to the request variables / -headers but I need the body of the request. Maybe there is a better way to get that - if so I didn't find it ;-)

nickvergessen replied:
Or you post a json body to the server using parameters instead of the body? ;)

fancycode replied:
You mean sending a body `body=the-json-object` and then using the value of variable `body`? That sounds hackish ;-)

nickvergessen commented on lib/Controller/BackendController.php:
```diff
@@ -0,0 +1,203 @@
+use OCP\IRequest;
+use OCP\Security\ISecureRandom;
+
+class BackendController extends Controller {
```
This controller has no routes, drop it?
Or you forgot to commit something?

fancycode replied:
No longer a controller now.

nickvergessen commented on lib/Controller/BackendController.php:
```diff
@@ -0,0 +1,203 @@
+	 * @param Room $room
+	 * @param array $userIds
+	 */
+	public function roomInvited($room, $userIds) {
```
this is only reachable for admins, right?

nickvergessen commented:
The main question is at which point the things should be decoupled.
I guess currently this app contains the bridge code as well, right?
So that the "external" signalings can focus on signaling instead of also having to do the php app code to listen to the events, etc?

nickvergessen commented on lib/Room.php:
```diff
@@ -214,34 +241,54 @@ public function changeType($newType) {
+				'type' => $participantType,
+				'sessionId' => $sessionId,
+			]));
+		foreach ($participants as &$participant) {
```
remove the reference `&`

fancycode replied:
Done

nickvergessen commented on lib/Room.php:
```diff
@@ -214,34 +241,54 @@ public function changeType($newType) {
+				'sessionId' => $sessionId,
+			]));
+		foreach ($participants as &$participant) {
+			$query = $this->db->getQueryBuilder();
```
Please create this outside of the `foreach`, use `$query->createParameter('userId')` and so on, and inside the loop just call `setParameter('userId', $participant)->setParameter(....); $query->execute();`

fancycode replied:
Changed

nickvergessen commented on lib/AppInfo/Application.php:
```diff
@@ -29,4 +30,29 @@ public function __construct(array $urlParams = []) {
+		$notifier = $this->getContainer()->query(BackendNotifier::class);
+
+		$dispatcher = $this->getContainer()->getServer()->getEventDispatcher();
+		$dispatcher->addListener('\OCA\Spreed\Room::postAddParticipants', function($event) use ($notifier) {
```
I'd use `->addListener(Room::class . '::postAddParticipants')` like we do in other places as well already

fancycode replied:
Changed. I based my code on the `dav` app, where the complete strings are used - using the classes makes more sense though!

nickvergessen commented on lib/Room.php:
```diff
@@ -146,6 +152,8 @@ public function getParticipant($userId) {
 	}
 
 	public function deleteRoom() {
+		$this->dispatcher->dispatch('\OCA\Spreed\Room::preDeleteRoom',
```
`self::class . '::preDeleteRoom'`

fancycode replied:
Changed

fancycode commented:
Rebased to current master.

nickvergessen commented:
Rebased this, the "pre rebase" version is backed up in the https://github.com/nextcloud/spreed/tree/standalone-signaling-before-rebase branch

nickvergessen commented:
Todo:
- [ ] Bring back multi-signaling-server option: Q: Do they all use the same secret @fancycode 
- [ ] Check if more events need to be listened to

What else is open?

nickvergessen commented:
Ping for the check of events

fancycode commented:
Rebased and fixed various problems with the new configuration code ;-) Now checking new events...

I'll also think about how at least some parts could be automatically tested to prevent future breakage due to seemingly unrelated changes.

fancycode commented:
@nickvergessen This should be good for another (and hopefully final) review now.

fancycode commented:
Rebased again.

nickvergessen commented:
Linting fail is unrelated

nickvergessen commented:
Only increased the version to trigger update of the settings again.

Merging
